### PR TITLE
devmapper: Log start and end of filesystem creation

### LINKED ISSUE #16653: docker timeouts when waiting for mkfs to finish on devicemapper storage driver device
# Description
```
docker version: docker-1.7.1-115.el7.x86_64
docker info: N/A
uname -a: Linux test-node-compute-f429d.example.com 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
platform: libvirt (kvm)
```

By default on RHEL 7.1 docker creates sparse file one upon start (systemctl start docker). That file is used as a loopback-device (seemingly using dmsetup syscall eqivalents), then issues 

    mkfs.ext4 -E nodiscard,lazy_itable_init=0,lazy_journal_init=0 /dev/mapper/docker-253:1-243367-base

This command may take several minutes to complete. Meanwhile the ```avgqu-sz``` gauge of sar shoots up to >25000. Then the systemctl start reports  timeout.

# Reproducer
Remove the backing files. I am not sure about the order of dmsetup lines here.
```
# systemctl stop docker
#dmsetup remove /dev/dm-0
#dmsetup remove /dev/dm-1
# rm /var/lib/docker/devicemapper/devicemapper/*
# systemctl start docker
```
... BLOCKING

```
# sar -d 1
...
15:28:24          DEV       tps  rd_sec/s  wr_sec/s  avgrq-sz  avgqu-sz     await     svctm     %util
15:28:25     dev253-0      0,00      0,00      0,00      0,00      0,00      0,00      0,00      0,00
15:28:25    dev253-16      0,00      0,00      0,00      0,00      0,00      0,00      0,00      0,00
15:28:25     dev252-0      0,00      0,00      0,00      0,00    112,86      0,00      0,00    100,40
15:28:25     dev252-1      0,00      0,00      0,00      0,00  34883,39      0,00      0,00    100,40
```

For some time, we can see the mkfs being performed
```
#ps aux | grep mkfs
mkfs.ext4 -E nodiscard,lazy_itable_init=0,lazy_journal_init=0 /dev/mapper/docker-253:1-243367-base
```

But the user-space process disappears before the kernel-space process, while the queue is still quite full.

The result is 
```
[root@test-node-compute-f429d openshift]# systemctl status -l docker.service
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled)
  Drop-In: /usr/lib/systemd/system/docker.service.d
           └─docker-sdn-ovs.conf
   Active: failed (Result: timeout) since Út 2015-09-29 15:30:25 UTC; 1min 53s ago
     Docs: http://docs.docker.com
 Main PID: 25025

zář 29 15:27:25 test-node-compute-f429d.example.com systemd[1]: Starting Docker Application Container Engine...
zář 29 15:27:25 test-node-compute-f429d.example.com docker[25025]: time="2015-09-29T15:27:25.975880434Z" level=warning msg="--storage-opt dm.thinpooldev is preferred over --storage-opt dm.datadev or dm.metadatadev"
zář 29 15:27:25 test-node-compute-f429d.example.com docker[25025]: time="2015-09-29T15:27:25.976662822Z" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)"
zář 29 15:27:25 test-node-compute-f429d.example.com docker[25025]: time="2015-09-29T15:27:25.978128795Z" level=error msg="Unable to delete device: Error running DeleteDevice dm_task_run failed"
zář 29 15:28:55 test-node-compute-f429d.example.com systemd[1]: docker.service operation timed out. Terminating.
zář 29 15:30:25 test-node-compute-f429d.example.com systemd[1]: docker.service stopping timed out (2). Killing.
zář 29 15:30:25 test-node-compute-f429d.example.com systemd[1]: Failed to start Docker Application Container Engine.
zář 29 15:30:25 test-node-compute-f429d.example.com systemd[1]: Unit docker.service entered failed state.
```

It seems mkfs wrote about at least 150MB of data:
```
#du -h /var/lib/docker/devicemapper/devicemapper/*
150M	/var/lib/docker/devicemapper/devicemapper/data
672K	/var/lib/docker/devicemapper/devicemapper/metadata
```
subsequent mkfs executed on the device exits fast, I think this is caused by the fact it is thin-provisioned.

I can't see any meaningfull error message in the logs that would help me debugging this problem.

Note that when one tries to start the docker again, docker will re-create the sparse file so the start is likely to fail again even though the FS is created fine every time, because this is done asychronously

# Expected state
Docker started on system no matter how much loaded it is. This would probably need checking of the mkfs progress.

Maybe the -D mkfs option may help.

       -D     Use  direct I/O when writing to the disk.  This avoids mke2fs dirtying a lot of
              buffer cache memory, which may impact other  applications  running  on  a  busy
              server.   This  option  will  cause mke2fs to run much more slowly, however, so
              there is a tradeoff to using direct I/O.

### PR DESCRIPTION
**Pull request by rhvgoyal**
Fixes #16653 

ext4 filesystem creation can take a long time on 100G thin device and
systemd might time out and kill docker service. Often user is left thinking
why docker is taking so long and logs don't give any hint. Log an info
message in journal for start and end of filesystem creation. That way
a user can look at logs and figure out that filesystem creation is
taking long time.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

### CODE REVIEW

File: daemon/graphdriver/devmapper/deviceset.go [hunk 1]
```diff
@@ -590,6 +590,9 @@ func (devices *DeviceSet) createFilesystem(info *devInfo) error {
 		devices.filesystem = determineDefaultFS()
 	}
 
+	logrus.Infof("devmapper: Creating filesystem %s on device(%s). START", devices.filesystem, info.Name())
+	defer logrus.Infof("devmapper: Creating filesystem %s on device(%s). END", devices.filesystem, info.Name())
```
Comment (estesp) [inline daemon/graphdriver/devmapper/deviceset.go]: potentially minor, but with a `defer` executed log entry, that means that even in all error cases you get the log entry, even if no actual filesystem creation happened?  Overall this seems like a good thing to log, but wonder if it should be more clearly aligned with actual creation starting and finishing..

Comment (rhvgoyal) [inline daemon/graphdriver/devmapper/deviceset.go]: @estesp right and that's why we never claim that filesystem creation successfully finished. We just say START and END in an attempt to figure out when does operation finish (success/failure).

If personally feel that we are already doing this in many of the functions where we are logging START/END of function and END does not mean success. It just means that you returned from function.

So I feel this is good enough. But if you feel strongly about it, I can change it.

**Comment by calavera** [comment]
I agree with @estesp that the exit logging can be confusing. I think we should be explicit if we're going to add this information. At exit, we should say whether this operation finished successfully or not.

**Comment by rhvgoyal** [comment]
@calavera @estesp Ok, I have pushed another version of patch which will explicitly say whether filesystem creation succeeded or not. PTAL.

**Comment by calavera** [comment]
I know it's not part of this change, but, can we replace the 4 lines at the bottom of the function with a single `return`?

This code:

```go
 	if err != nil {
 		return err
 	}
 
 	return nil
```

Can be:

```go
    return
```

Regardless, LGTM.

**Comment by rhvgoyal** [comment]
yep we can do that. will do it tomorrow.

**Comment by rhvgoyal** [comment]
@calavera made the changes as you had mentioned.

**Comment by estesp** [comment]
LGTM

**Pull request merged**